services/repository/generate.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package repository
+
+import (
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/notification"
+)
+
+// GenerateRepository generates a repository from a template
+func GenerateRepository(doer, owner *models.User, templateRepo *models.Repository, opts models.GenerateRepoOptions) (*models.Repository, error) {
+	generateRepo, err := models.GenerateRepository(doer, owner, templateRepo, opts)
+	if err != nil {
+		if generateRepo != nil {
+			if errDelete := models.DeleteRepository(doer, owner.ID, generateRepo.ID); errDelete != nil {
+				log.Error("Rollback deleteRepository: %v", errDelete)
+			}
+		}
+		return nil, err
+	}
+
+	ctx, sess, err := models.TxDBContext()
```

Comment:
I'm not saying this is incorrect, but why is this a different transaction from the `models.GenerateRepository()` above? Can't that function take a DBContext? Or is it that you preferred to generate as much as possible of the repo even if you get an error at some point later?

Reply:
The repo needs to be created so that later transactions have access to its ID.

Reply:
As for `GenerateRepository` taking a context, it could but there was no need to pass back a session so I didn't think it was necessary.

Reply:
@jolheiser AFAIK you get the ID when the repo row is inserted (Magic by Xorm™), so you have access to the ID before commit. That's how the issue.index setting works.

Reply:
Interesting, I would have thought ID's come back when transactions are committed.  
If that's not true then it could be the same session, I suppose.

Reply:
You were right, black magic prevails. The entire repo generation should now use a single session.  
It is only committed before notifications because that uses another session internally.